src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -111,35 +107,63 @@ namespace ts.refactor.convertToOptionalChainExpression {
     function getOccurrencesInExpression(matchTo: Expression, expression: Expression, checker: TypeChecker): (PropertyAccessExpression | Identifier)[] | undefined {
         const occurrences: (PropertyAccessExpression | Identifier)[] = [];
         while (isBinaryExpression(expression) && expression.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken) {
-            const match = checkMatch(matchTo, expression.right, checker);
+            const match = getMatchingSubexpression(matchTo, expression.right, checker);
             if (!match) {
                 break;
             }
             occurrences.push(match);
             matchTo = match;
             expression = expression.left;
         }
-        const finalMatch = checkMatch(matchTo, expression, checker);
+        const finalMatch = getMatchingSubexpression(matchTo, expression, checker);
         if (finalMatch) {
             occurrences.push(finalMatch);
         }
         return occurrences.length > 0 ? occurrences: undefined;
     }
 
     /**
-     * Checks that expression is a subchain of matchTo.
+     * Checks that expression is a syntactic subexpression of matchTo.
      */
-    function checkMatch(matchTo: Expression, expression: Expression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
+    function getMatchingSubexpression(matchTo: Expression, expression: Expression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
         if (isCallExpression(matchTo)) {
-            // Recurse through the call expressions to match a.b to a.b()().
-            return !isCallExpression(expression) ? checkMatch(matchTo.expression, expression, checker) : undefined;
+            return !isCallExpression(expression) ? getMatchingSubexpression(matchTo.expression, expression, checker) : undefined;
         }
         else if ((isPropertyAccessExpression(expression) || isIdentifier(expression)) && (isPropertyAccessExpression(matchTo) || isIdentifier(matchTo))) {
-            return checker.isOrContainsMatchingReference(matchTo, expression) ? expression : undefined;
+            return containsSyntacticSubchain(matchTo, expression, checker) ? expression : undefined;
         }
         return undefined;
     }
 
+    /**
+     * Returns true if target is a syntactic subchain of source.
+     */
+    function containsSyntacticSubchain(source: Node, target: Node, checker: TypeChecker): boolean {

```

Comment:
And maybe `chainStartsWith(chain, subchain)` is a clearer name for the function, since IIUC the subchain (currently called `target`) has to appear at the _beginning_ of the chain (currently `source`), not just anywhere within it.

Reply:
Good point! Updated.